turbo/stages/headerdownload/header_algos.go

```diff
@@ -682,6 +719,75 @@ func (hd *HeaderDownload) InsertHeaders(hf func(header *types.Header, headerRaw
 	return hd.highestInDb >= hd.preverifiedHeight && hd.topSeenHeight > 0 && hd.highestInDb >= hd.topSeenHeight, nil
 }
 
+func (hd *HeaderDownload) InsertHeadersBackwards(tx kv.RwTx, canonicalHashCollector *etl.Collector, headerCollector *etl.Collector, logPrefix string, logChannel <-chan time.Time) (bool, error) {
+	if len(hd.insertList) == 0 {
+		return false, nil
+	}
+	hd.lock.Lock()
+	defer hd.lock.Unlock()
+	sort.Slice(hd.insertList, func(i, j int) bool {
+		return hd.insertList[i].header.Number.Uint64() > hd.insertList[j].header.Number.Uint64()
+	})
+	for len(hd.insertList) > 0 {
+		header := hd.insertList[0].header
+		if header.Number.Uint64() >= hd.lastProcessedPayload {
+			hd.insertList = hd.insertList[1:]
+			continue
+		}
+		// If we miss some headers or an header results to be invalid, we ask for them again
+		if header.Number.Uint64() != hd.lastProcessedPayload-1 {
+			break
+		}
+		if header.Hash() != hd.expectedHash {
+			hd.insertList = hd.insertList[1:]
```

Comment:
If the hash does not match, you don't really want to keep anything in the `insertList`, so approach I suggested above would work